src/qcodes/dataset/database_extract_runs.py

```diff
@@ -168,3 +175,229 @@ def _extract_single_dataset_into_db(
+        
+        # Get all run IDs from the source database
+        run_ids = get_runs(source_conn)
+        log.info(f"Found {len(run_ids)} datasets to process")
+        
+        if not run_ids:
+            log.warning("No datasets found in source database")
+            return {}
+        
+        # Process datasets by experiment to preserve structure
+        result_status = {}
+        processed_experiments = {}  # Map source exp_id to target exp_id
+        
+        for run_id in run_ids:
+            try:
+                dataset = DataSet(run_id=run_id, conn=source_conn)
+                exp_id = dataset.exp_id
+                
+                # Create experiment in target DB if not already done
+                if exp_id not in processed_experiments:
+                    exp_attrs = get_experiment_attributes_by_exp_id(source_conn, exp_id)
+                    
+                    with atomic(target_conn) as target_conn_atomic:
+                        target_exp_id = _create_exp_if_needed(
+                            target_conn_atomic,
+                            exp_attrs["name"],
+                            exp_attrs["sample_name"],
+                            exp_attrs["format_string"],
+                            exp_attrs["start_time"],
+                            exp_attrs["end_time"],
+                        )
+                    processed_experiments[exp_id] = target_exp_id
+                    log.info(f"Created experiment '{exp_attrs['name']}' in target database")
+                else:
+                    target_exp_id = processed_experiments[exp_id]
+                
+                # Try to export dataset to NetCDF and create metadata-only version
+                status = _process_single_dataset(
+                    dataset, source_conn, target_conn, export_path, target_exp_id
+                )
+                result_status[run_id] = status
+                
+            except Exception as e:
+                log.error(f"Failed to process dataset {run_id}: {e}")
+                result_status[run_id] = f"failed: {str(e)}"
+        
+        log.info(f"Processing complete. Status summary: {result_status}")
+        return result_status
+        
+    except Exception as e:
+        log.error(f"Database operation failed: {e}")
+        raise
+    finally:
+        if source_conn is not None:
+            source_conn.close()
+        if target_conn is not None:
+            target_conn.close()
+
+
+def _process_single_dataset(
+    dataset: DataSet,
+    source_conn: AtomicConnection,
+    target_conn: AtomicConnection,
+    export_path: Path,
+    target_exp_id: int,
+) -> str:
+    """
+    Process a single dataset: export to NetCDF and create metadata-only version
+    or copy as-is if export fails.
+    
+    Returns:
+        Status string indicating what was done with the dataset
+    """
+    run_id = dataset.run_id
+    
+    # Check if dataset is already in target database
```

Comment:
it's a good idea to check if the dataset has already been moved to the target database file, but please do it by checking against the GUID of the dataset as those are guaranteed to be unique, while run_id's are not.